### PR TITLE
Add horizontal swing to climate

### PR DESCRIPTION
**Pull request by gjohansson-ST**
## Proposed change

Add horizontal swing mode to `ClimateEntity`
Core PR: https://github.com/home-assistant/core/pull/125578

### Example printscreens

image
With both `swing` and `horizontal_swing` it becomes vertical and horizontal.
image
If only `swing` it remains as swing
image

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

## Release Notes

- **New Features**
	- Enhanced thermostat card with updated swing modes and support for horizontal swing settings.
	- Introduced new configuration options for climate control, including horizontal swing modes.
	- Added a user-friendly interface for selecting and customizing horizontal swing modes in climate settings.

- **Bug Fixes**
	- Improved handling of unavailable states for climate controls.

- **Documentation**
	- Added translations and labels for new horizontal swing mode features.

### CODE REVIEW

**Comment by matthiasdebaat** [comment]
When we introduce Horizontal swing mode, does that make the other one Vertical swing mode? To make the label shorter, we could drop the word mode: `Horizontal swing` and `Vertical swing`. I wouldn't drop swing, as that gives context about the feature.

**Comment by gjohansson-ST** [comment]
> When we introduce Horizontal swing mode, does that make the other one Vertical swing mode?

Yes

> To make the label shorter, we could drop the word mode: `Horizontal swing` and `Vertical swing`. I wouldn't drop swing, as that gives context about the feature.

I think that makes sense.
So when both swing modes are available they will be `Vertical swing` and `Horizontal swing`. When only one is available (it can be combined both vertical and horizontal or only one of them) it will be called `Swing`

Would it make sense to update `fan`, `preset` etc. in a separate PR then to align the language (or we keep `mode` here)?

**Comment by matthiasdebaat** [comment]
> Would it make sense to update `fan`, `preset` etc. in a separate PR then to align the language (or we keep `mode` here)?

I like that, I don't see any concerns if we drop mode. This also gives us one option with Mode:
CleanShot 2024-10-29 at 14 27 35@2x

**Comment by gjohansson-ST** [comment]
> I like that, I don't see any concerns if we drop mode. This also gives us one option with Mode

I have introduced it with this PR but maybe should be done in a preliminary PR before introduced here.
The screenshots  in the PR description are updated with the new strings.

File: src/dialogs/more-info/controls/more-info-climate.ts [hunk 1]
```diff
@@ -197,10 +201,7 @@ class MoreInfoClimate extends LitElement {
         ${supportPresetMode && stateObj.attributes.preset_modes
           ? html`
               <ha-control-select-menu
-                .label=${this.hass.formatEntityAttributeName(
-                  stateObj,
-                  "preset_mode"
-                )}
+                .label=${this.hass.localize("ui.card.climate.preset")}
```
Comment (bramkragten) [inline src/dialogs/more-info/controls/more-info-climate.ts:204]: Why was this changed?

File: src/dialogs/more-info/controls/more-info-climate.ts [hunk 2]
```diff
@@ -296,10 +294,9 @@ class MoreInfoClimate extends LitElement {
         ${supportSwingMode && stateObj.attributes.swing_modes
           ? html`
               <ha-control-select-menu
-                .label=${this.hass.formatEntityAttributeName(
-                  stateObj,
-                  "swing_mode"
-                )}
+                .label=${supportSwingHorizontalMode
+                  ? this.hass.localize("ui.card.climate.vertical_swing")
+                  : this.hass.localize("ui.card.climate.swing")}
```
Comment (bramkragten) [inline src/dialogs/more-info/controls/more-info-climate.ts:299]: This should still use `formatEntityAttributeName`

File: src/dialogs/more-info/controls/more-info-climate.ts [hunk 3]
```diff
@@ -344,6 +341,56 @@ class MoreInfoClimate extends LitElement {
               </ha-control-select-menu>
             `
           : nothing}
+        ${supportSwingHorizontalMode &&
+        stateObj.attributes.swing_horizontal_modes
+          ? html`
+              <ha-control-select-menu
+                .label=${this.hass.localize("ui.card.climate.horizontal_swing")}
```
Comment (bramkragten) [inline src/dialogs/more-info/controls/more-info-climate.ts:348]: Should use `formatEntityAttributeName`

File: src/translations/en.json [hunk 4]
```diff
@@ -125,6 +125,10 @@
         "high": "high",
         "low": "low",
         "mode": "Mode",
+        "fan": "Fan",
+        "swing": "Swing",
+        "vertical_swing": "Vertical swing",
+        "horizontal_swing": "Horizontal swing",
```
Comment (bramkragten) [inline src/translations/en.json:131]: These translations are provided by the backend, they should not be duplicated here

File: src/dialogs/more-info/controls/more-info-climate.ts (hunk 1 above)
Comment (gjohansson-ST) [reply src/dialogs/more-info/controls/more-info-climate.ts:204]: See my response [here](https://github.com/home-assistant/frontend/pull/22043#issuecomment-2453580647).
It's related to remove the word "mode" from the discussion with Matthias above. So in this case the label becomes `Preset` instead of `Preset mode`.

File: src/dialogs/more-info/controls/more-info-climate.ts [hunk 5]
```diff
@@ -248,10 +249,7 @@ class MoreInfoClimate extends LitElement {
         ${supportFanMode && stateObj.attributes.fan_modes
           ? html`
               <ha-control-select-menu
-                .label=${this.hass.formatEntityAttributeName(
-                  stateObj,
-                  "fan_mode"
-                )}
+                .label=${this.hass.localize("ui.card.climate.fan")}
```
[the remark this answers is not in the document]
Comment (gjohansson-ST) [reply src/dialogs/more-info/controls/more-info-climate.ts:252]: Same answer as previous

File: src/dialogs/more-info/controls/more-info-climate.ts (hunk 1 above)
Comment (bramkragten) [reply src/dialogs/more-info/controls/more-info-climate.ts:204]: Then that should be updated in core, not in frontend? It also seems like a change that can be done in a separate PR?

Comment (gjohansson-ST) [reply src/dialogs/more-info/controls/more-info-climate.ts:204]: I think that was a bit the point to disconnect the backend state attributes with the frontend translations but ok, let's revert that back to current and let's do/discuss it in a separate PR.

Renaming those state attributes in the backend would be a major overhaul which I don't think we would want.

Comment (gjohansson-ST) [reply src/dialogs/more-info/controls/more-info-climate.ts]: All the related translation changes has been removed from this PR.

**Review by bramkragten — APPROVED** [review]

**Review by jwillemsen — APPROVED** [review]

**Comment by apppie123** [comment]
I just saw this PR and I have a command/questing regarding the names.
I see it has been decided to call it "horizontal swing" and "vertical swing".

However, "swing" is actually a mode with for example Mitsubishi Heavy Industries and Mitsubishi Electric aircons.
They call these "Horizontal Vanes" and "Vertical Vanes" or "Louvers" or "flaps".

It's most common to call them vanes in they DIY HVAC module world.

Calling them swings and adding a "swing" mode might cause confusion. Valid modes for above two brands are 1, 2, 3, 4, auto, swing. While they also offer a horizontal vane 'split left-right'.

**Comment by janmashat** [comment]
It's currently called "[Swing mode](https://developers.home-assistant.io/docs/core/entity/climate/#swing-modes)" - so changing that to "Vane mode" would be confusing IMO...

Screenshot 2024-11-30 at 11 52 42

However, I agree that `Vertical vane` and `Horizontal vane` would be better labels when both modes are available (and the latter might even fit on the screen) - especially since not all of the values will involve swinging, for example SwingMiddle vs FixedMiddle (in the case of [Gree](https://github.com/cmroche/greeclimate/blob/master/greeclimate/device.py#L72-L96)).

**Comment by apppie123** [comment]
Yes I agree. I should have rephrased it a bit.
The current "swing mode" should stay as legacy(and used for very basic air conditioners with only these options). I can imagine it was called swing mode in the past when there was only vertical, horizontal or both swinging of the vanes. It makes sense, because the unit would only be able "swing" around.

But in case of modern air conditioners it is about the absolute position of the vanes/louvers.
Vertical Vane and Horizontal Vane makes much more sense to me and there are many fixed positions to choose from.

**Comment by karwosts** [comment]
Just FYI if anyone tries to use this, the UI for this is broken in 2024.12.0, missed an import

See https://github.com/home-assistant/frontend/pull/23168
